platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractProperty.java

```diff
@@ -234,10 +231,12 @@ protected void beforeRead(ValueConsumer consumer) {
         beforeRead(producer, consumer);
     }
 
+    protected void beforeReadNoProducer(ValueConsumer consumer) {
+        beforeRead(null, consumer);
+    }
+
     private void beforeRead(@Nullable ModelObject effectiveProducer, ValueConsumer consumer) {
-        if (state.maybeFinalizeOnRead(getDisplayName(), effectiveProducer, consumer)) {
-            finalizeNow(state.forUpstream(consumer));
-        }
+        state.finalizeOnReadIfNeeded(this::getDisplayName, effectiveProducer, consumer, this::finalizeNow);
```

Comment:
`finalizeOnReadIfNeeded()`  implements conditional finalization now (so it can be reused)